device/simulation/tt_simulation_device.cpp

```diff
@@ -0,0 +1,261 @@
+    std::memcpy(mem_ptr, rd_resp.data(), rd_resp.size() * sizeof(uint32_t));
+}
+
+void tt_SimulationDevice::read_from_device(std::vector<uint32_t> &vec, tt_cxy_pair core, uint64_t addr, uint32_t size, const std::string& tlb_to_use) {
+    void *rd_resp;
+    size_t rd_rsp_sz;
+
+    // Send read request
+    auto rd_req_buf = create_flatbuffer(DEVICE_COMMAND_READ, {0}, core, addr, size);
+    host.send_to_device(rd_req_buf.GetBufferPointer(), rd_req_buf.GetSize());
+
+    // Get read response
+    host.recv_from_device(&rd_resp, &rd_rsp_sz);
+
+    auto rd_resp_buf = GetDeviceRequestResponse(rd_resp);
+    if (addr != 0x40){
+        log_info(tt::LogEmulationDriver, "Device reading vec");
+        print_flatbuffer(rd_resp_buf); // 0x40 is host polling device, don't print since it'll spam
+    }
+    std::vector<uint32_t> rd_resp_data(rd_resp_buf->data()->begin(), rd_resp_buf->data()->end());
+
+    vec = rd_resp_data;
+    nng_free(rd_resp, rd_rsp_sz);
+}
+
+// For fd cqs
+void tt_SimulationDevice::write_to_sysmem(std::vector<uint32_t>& vec, uint64_t addr, uint16_t channel, chip_id_t src_device_id) {}
+void tt_SimulationDevice::write_to_sysmem(const void* mem_ptr, std::uint32_t size,  uint64_t addr, uint16_t channel, chip_id_t src_device_id) {}
+void tt_SimulationDevice::read_from_sysmem(std::vector<uint32_t> &vec, uint64_t addr, uint16_t channel, uint32_t size, chip_id_t src_device_id) {}
+void tt_SimulationDevice::read_from_sysmem(void* mem_ptr, uint64_t addr, uint16_t channel, uint32_t size, chip_id_t src_device_id) {}
+
+void tt_SimulationDevice::wait_for_non_mmio_flush() {}
+void tt_SimulationDevice::l1_membar(const chip_id_t chip, const std::string& fallback_tlb, const std::unordered_set<tt_xy_pair>& cores) {}
+void tt_SimulationDevice::dram_membar(const chip_id_t chip, const std::string& fallback_tlb, const std::unordered_set<uint32_t>& channels) {}
+void tt_SimulationDevice::dram_membar(const chip_id_t chip, const std::string& fallback_tlb, const std::unordered_set<tt_xy_pair>& cores) {}
+
+// Misc. Functions to Query/Set Device State
+std::unordered_map<chip_id_t, uint32_t> tt_SimulationDevice::get_harvesting_masks_for_soc_descriptors() {
+    return {{0, 0}};
+}
+
+// static int tt_SimulationDevice::detect_number_of_chips() {}
+
+std::vector<chip_id_t> tt_SimulationDevice::detect_available_device_ids() {
+    return {0};
+}
+// std::set<chip_id_t> tt_SimulationDevice::get_target_mmio_device_ids() {
+//     return std::set<chip_id_t>();
+// }
+
+std::set<chip_id_t> tt_SimulationDevice::get_target_remote_device_ids() {
+    return target_remote_chips;
+}
+
+std::map<int,int> tt_SimulationDevice::get_clocks() {
+    return {{0, 0}};
+}
+
+void *tt_SimulationDevice::host_dma_address(std::uint64_t offset, chip_id_t src_device_id, uint16_t channel) const {
+    // might need to implement later
+    return nullptr;
+}
+
+std::uint64_t tt_SimulationDevice::get_pcie_base_addr_from_device() const {
```

Comment:
we could probably commonize this with tt_SiliconDevice?

Reply:
Wdym by commonize? Right now, I just copy-pasted what's in the `tt_SilliconDevice` to here

Reply:
I guess if one changes we would need to rmr to update this as well, if they pull from a common function we don't have to do that

Reply:
@joelsmithTT can I move this implementation as well these
```
tt_device_dram_address_params dram_address_params;
tt_device_l1_address_params l1_address_params;
tt_driver_host_address_params host_address_params;
tt_driver_eth_interface_params eth_interface_params;
```
into `tt_device.h` or should I leave this cleanup work for later?

Reply:
Moving all to the base class would be an improvement.  It's up to you if you want to do it in this branch or in future work.  Maybe create an issue for it if you decide on the latter.

Reply:
https://github.com/tenstorrent/tt-umd/issues/26